execution-plane/arcon/src/allocator/mod.rs

```diff
@@ -45,17 +44,25 @@ impl ArconAllocator {
             return arcon_err!("{}", "Cannot alloc for 0 sized pointer");
         }
         let (size, align) = (std::mem::size_of::<T>(), std::mem::align_of::<T>());
-        let required_bytes = size * capacity;
+
+        let required_bytes = capacity
+            .checked_mul(size)
+            .ok_or(arcon_err_kind!("{}", "Capacity overflow"))?;
 
         if self.curr_alloc + required_bytes > self.limit {
-            return arcon_err!("{}", "OutOfMemory");
+            return arcon_err!("{}", "ArconAllocator is OOM");
         }
 
         let layout = Layout::from_size_align_unchecked(required_bytes, align);
         let mem = System.alloc(layout);
+
+        if mem.is_null() {
+            return arcon_err!("{}", "SystemAllocator is OOM");
+        }
+
         self.curr_alloc += layout.size();
+        let id = self.alloc_counter;
         self.alloc_counter += 1;
```

Comment:
This could theoretically overflow in a very long running streaming pipeline with many allocations and deallocations. Should be more specific as to the failure behaviour?

Reply:
How about AllocId = (u64,u64) where (epoch, id) and id reaches u64::MAX_SIZE, then we simply bump the alloc epoch and start from 0 in id again?

Reply:
Works. We could also try to keep track of the lowest active allocation and wrap around if it's ok to do so and error otherwise. It really depends on how many active allocations we want to support and how likely we think it is that they are long lived.
I don't really have a particular opinion on the matter. I just wanted to be sure that we are not silently ignoring an overflow, since in release Rust they don't generate a panic.

Reply:
`u128` is also a thing in Rust :)

Reply:
That, too^^